### DISCUSSION TITLE
is there any way to use for Chrome-Extension-of-Manifest-V3？

### DISCUSSION THREAD

**Discussion by Achi-Z** [ujjwalguptaofficial/JsStore#250] [Q&A] [answered]
is there any way to use for Chrome-Extension-of-Manifest-V3？

**Comment by ujjwalguptaofficial** [comment, marked as answer]
I don't have experience of extension development, so no idea.

**Comment by reedjones** [comment]
also wondering, if anyone knows?

**Reply by ujjwalguptaofficial** [reply]
I have not done any POC, but chatgpt says we can use. Here is the response - 

Yes, you can use IndexedDB in a Chrome extension! IndexedDB is a low-level API for storing large amounts of structured data, and it’s supported in Chrome, so it works well within Chrome extensions.

Here’s a brief overview of how to use IndexedDB in a Chrome extension:

1. **Manifest File**: Make sure your Chrome extension's `manifest.json` file has the appropriate permissions if you need to access web content or interact with specific domains. For example:
    ```json
    {
      "manifest_version": 3,
      "name": "My Extension",
      "version": "1.0",
      "permissions": [
        "storage",
        "activeTab"
      ],
      "background": {
        "service_worker": "background.js"
      },
      "content_scripts": [
        {
          "matches": ["<all_urls>"],
          "js": ["content.js"]
        }
      ]
    }
    ```

2. **Background Script**: If you’re using a background script (`background.js`), you can use IndexedDB directly there to store or retrieve data:
    ```javascript
    const dbName = 'myDatabase';
    const dbVersion = 1;

    function openDatabase() {
      const request = indexedDB.open(dbName, dbVersion);

      request.onupgradeneeded = function(event) {
        const db = event.target.result;
        if (!db.objectStoreNames.contains('myObjectStore')) {
          db.createObjectStore('myObjectStore', { keyPath: 'id', autoIncrement: true });
        }
      };

      request.onsuccess = function(event) {
        console.log('Database opened successfully');
        const db = event.target.result;
        // Perform database operations here
      };

      request.onerror = function(event) {
        console.error('Database error:', event.target.errorCode);
      };
    }

    openDatabase();
    ```

3. **Content Script**: If you're running code in a content script (`content.js`), you can also use IndexedDB to interact with web pages. The same IndexedDB API methods apply.

4. **Error Handling**: Ensure you handle errors and manage database versions properly. You can use the `onerror` and `onsuccess` event handlers to manage the outcomes of your database operations.

5. **Chrome Extension Specifics**: Remember that your extension's scripts (background, content) have different contexts. Data saved in IndexedDB in a content script won't be accessible directly from a background script or popup script unless you use messaging to synchronize data.

Using IndexedDB in a Chrome extension is very similar to using it in a regular web application, with the added flexibility of Chrome's extension architecture.

**Reply by ujjwalguptaofficial** [reply]
which means we can also use jsstore.

**Reply by reedjones** [reply]
> Using IndexedDB in a Chrome extension is very similar to using it in a regular web application, with the added flexibility of Chrome's extension architecture.

yeah that code doesn't work
and the statement above is *kinda* true I guess - but no not really.

I've been trying to get it working in an extension with no success. 
one key problem has to do with how to initiate JsStore with the worker url and so on, will be different in chrome extensions 
can you create a worker from within a worker in chrome background script?  should the worker script should be included as web accessible resource?  etc.. etc...

**Reply by ujjwalguptaofficial** [reply]
if worker url does not work, then you can make it work without worker also. 

check this doc - https://jsstore.net/docs/connection#without-web-worker

**Comment by dd3v** [comment]
I use JsStore in my Chrome extension. If it's still relevant, feel free to check it out

https://github.com/dd3v/favbox

**Reply by ujjwalguptaofficial** [reply]
Thanks @dd3v , that looks great project.
